File: src/engine/recommender.py
from typing import List, Dict, Any, Optional
from .models import Venue

class RecommenderEngine:
    """
    Thuật toán lọc và xếp hạng địa điểm đi chơi/cà phê theo khung giờ và sở thích.
    """

    @staticmethod
    def filter_and_rank_venues(
        venues: List[Venue],
        slot_tag: str,
        group_size: int = 1,
        preferences: Optional[List[str]] = None,
        max_price: Optional[str] = None,
        category: Optional[str] = None,
        city: Optional[str] = None,
        ward: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Lọc cứng (Hard Filter) và Phân điểm (Scoring) danh sách địa điểm.
        """
        if preferences is None:
            preferences = []

        scored_venues = []

        for venue in venues:
            # 1. Hard Filter: Time tag match (Ngoại trừ trường hợp venue phù hợp mọi khung giờ "all")
            if slot_tag not in venue.time_tags and "all" not in venue.time_tags:
                continue

            # 2. Hard Filter: Sức chứa vừa nhóm
            if venue.capacity < group_size:
                continue

            # 3. Hard Filter: Category nếu được yêu cầu
            if category and category != "all" and venue.category != category:
                continue

            # 4. Hard Filter: City nếu được yêu cầu
            if city and city != "all" and venue.city.lower() != city.lower():
                continue

            # 5. Hard Filter: Ward (Xã/Phường) nếu được yêu cầu
            if ward and ward != "all" and venue.ward.lower() != ward.lower():
                continue

            # --- SOFT FILTER & SCORING ---
            score = 0.0

            # Tag preference matching bonus (+10 điểm mỗi tag trùng)
            pref_matches = 0
            if preferences:
                for pref in preferences:
                    if any(pref.lower() in t.lower() for t in venue.tags):
                        pref_matches += 1
                score += pref_matches * 10.0

            # Rating score (+ rating * 5 điểm)
            score += venue.rating * 5.0

            # Price preference bonus
            if max_price:
                if venue.price_range == max_price:
                    score += 5.0

            scored_venues.append({
                "venue": {
                    "id": venue.id,
                    "name": venue.name,
                    "category": venue.category,
                    "time_tags": venue.time_tags,
                    "price_range": venue.price_range,
                    "capacity": venue.capacity,
                    "address": venue.address,
                    "city": venue.city,
                    "ward": venue.ward,
                    "lat": venue.lat,
                    "lng": venue.lng,
                    "tags": venue.tags,
                    "rating": venue.rating,
                    "image_url": venue.image_url
                },
                "score": round(score, 2),
                "pref_matches": pref_matches
            })

        # Sort by score descending, then rating descending
        ranked = sorted(scored_venues, key=lambda x: (x["score"], x["venue"]["rating"]), reverse=True)
        return ranked

File: src/engine_ts/recommender.ts
import { Venue } from "./models";

export interface FilterOptions {
    slot_tag: string;
    group_size?: number;
    preferences?: string[];
    max_price?: string;
    category?: string;
    city?: string;
    ward?: string;
}

export interface RankedVenueResult {
    venue: Venue;
    score: number;
    pref_matches: number;
}

export class RecommenderEngine {
    /**
     * Lọc cứng (Hard Filter) và Phân điểm (Scoring) danh sách địa điểm.
     */
    static filterAndRankVenues(
        venues: Venue[],
        options: FilterOptions
    ): RankedVenueResult[] {
        const {
            slot_tag,
            group_size = 1,
            preferences = [],
            max_price,
            category,
            city,
            ward
        } = options;

        const scoredVenues: RankedVenueResult[] = [];

        for (const venue of venues) {
            // 1. Hard Filter: Time tag match
            const hasTimeTag =
                venue.time_tags.includes(slot_tag) || venue.time_tags.includes("all");
            if (!hasTimeTag) continue;

            // 2. Hard Filter: Sức chứa vừa nhóm
            if (venue.capacity < group_size) continue;

            // 3. Hard Filter: Category nếu được yêu cầu
            if (category && category !== "all" && venue.category !== category) continue;

            // 4. Hard Filter: City nếu được yêu cầu
            if (city && city !== "all" && venue.city.toLowerCase() !== city.toLowerCase()) continue;

            // 5. Hard Filter: Ward nếu được yêu cầu
            if (ward && ward !== "all" && venue.ward.toLowerCase() !== ward.toLowerCase()) continue;

            // --- SOFT FILTER & SCORING ---
            let score = 0.0;
            let prefMatches = 0;

            if (preferences && preferences.length > 0) {
                for (const pref of preferences) {
                    const prefLower = pref.toLowerCase();
                    if (venue.tags.some((t) => t.toLowerCase().includes(prefLower))) {
                        prefMatches += 1;
                    }
                }
                score += prefMatches * 10.0;
            }

            score += (venue.rating || 4.5) * 5.0;

            if (max_price && venue.price_range === max_price) {
                score += 5.0;
            }

            scoredVenues.push({
                venue,
                score: Number(score.toFixed(2)),
                pref_matches: prefMatches
            });
        }

        // Sort by score descending, then rating descending
        scoredVenues.sort((a, b) => {
            if (b.score !== a.score) {
                return b.score - a.score;
            }
            return b.venue.rating - a.venue.rating;
        });

        return scoredVenues;
    }
}
